stl/inc/ranges

```diff
@@ -9169,6 +9169,632 @@ namespace ranges {
+                    const bool _Overflow = _Multiply_with_overflow_check(
+                        static_cast<_Difference_type<_Const>>(_RANGES size(_STD get<_Index>(_Parent->_Bases))),
+                        _Scaled_size<_Index + 1>(), _Product);
+                    _STL_VERIFY(!_Overflow, "Scaled-sum cannot be represented by _Difference_type<false> (N4928 "
+                                            "[range.cartesian.iterator]/8).");
+                    return _Product;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL != 0 / _ITERATOR_DEBUG_LEVEL == 0 vvv
+                    return static_cast<_Difference_type<_Const>>(_RANGES size(_STD get<_Index>(_Parent->_Bases)))
+                         * _Scaled_size<_Index + 1>();
+#endif //_ITERATOR_DEBUG_LEVEL == 0
+                } else {
+                    return static_cast<_Difference_type<_Const>>(1);
+                }
+            }
+
+            template <size_t _Index, class _Tuple>
+            _NODISCARD constexpr _Difference_type<_Const> _Scaled_distance(const _Tuple& _Tpl) const {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _Difference_type<_Const> _Product{};
+                const bool _Overflow = _Multiply_with_overflow_check(
+                    static_cast<_Difference_type<_Const>>(_STD get<_Index>(_Current) - _STD get<_Index>(_Tpl)),
+                    _Scaled_size<_Index + 1>(), _Product);
+                _STL_VERIFY(!_Overflow, "Scaled-sum cannot be represented by _Difference_type<false> (N4928 "
+                                        "[range.cartesian.iterator]/8).");
+                return _Product;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL != 0 / _ITERATOR_DEBUG_LEVEL == 0 vvv
+                return static_cast<_Difference_type<_Const>>(_STD get<_Index>(_Current) - _STD get<_Index>(_Tpl))
+                     * _Scaled_size<_Index + 1>();
+#endif //_ITERATOR_DEBUG_LEVEL == 0
+            }
+
+            template <class _Tuple, size_t... _Indices>
+            _NODISCARD constexpr _Difference_type<_Const> _Scaled_sum(
+                const _Tuple& _Tpl, index_sequence<_Indices...>) const {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _Difference_type<_Const> _Sum{};
+                const bool _Overflow = (_Add_with_overflow_check(_Sum, _Scaled_distance<_Indices>(_Tpl), _Sum) || ...);
+                _STL_VERIFY(!_Overflow, "Scaled-sum cannot be represented by _Difference_type<false> (N4928 "
+                                        "[range.cartesian.iterator]/8).");
+                return _Sum;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL != 0 / _ITERATOR_DEBUG_LEVEL == 0 vvv
+                return (_Scaled_distance<_Indices>(_Tpl) + ...);
+#endif //_ITERATOR_DEBUG_LEVEL == 0
+            }
+
+            template <class _Tuple>
+            _NODISCARD constexpr _Difference_type<_Const> _Distance_from(const _Tuple& _Tpl) const {
+                return _Scaled_sum(_Tpl, make_index_sequence<1 + sizeof...(_Rest)>{});
+            }

```

Comment:
I was working on this at some point, and I think we can reduce the entire implementation of `_Distance_from()` to the following **(sans overflow checks)**:

```C++
template <size_t _Index, class _Tuple>
_NODISCARD constexpr difference_type _Distance_from_impl(const _Tuple& _Rhs_tuple) const noexcept {
	const difference_type _Curr_diff =
		static_cast<difference_type>(_STD get<_Index>(_Current) - _STD get<_Index>(_Rhs_tuple));

        if constexpr (_Index > 0) {
		const difference_type _Curr_size =
                        static_cast<difference_type>(_RANGES size(_STD get<_Index>(_Parent->_Base)));
                return _Curr_diff + _Curr_size * _Distance_from_impl<_Index - 1>(_Rhs_tuple);
        } else {
                return _Curr_diff;
        }


template <class _Tuple>
_NODISCARD constexpr difference_type _Distance_from(const _Tuple& _Rhs_tuple) const noexcept {
	// This recursive implementation is a mathematically simplified version of the distance-from
	// function as specified in N4928: [range.cartesian.iterator]/7. It can drastically reduce the
	// number of multiplication operations when the number of views is high.

	return _Distance_from_impl<sizeof...(_Rest)>(_Rhs_tuple);
}
```

This works by simplifying the expression represented by the `_Distance_from()` function. For instance, consider the case where `sizeof...(_Rest) == 2`. Let `diff_N` represent `static_cast<difference_type>(std::get<N>(_Current) - std::get<N>(_Rhs_tuple))` and `size_N` represent `static_cast<difference_type>(ranges::size(std::get<N>(_Parent->_Bases)))`. The returned `scaled-sum` value is calculated as follows:

```
scaled-sum = scaled-distance(0) + scaled-distance(1) + scaled-distance(2)
           = (diff_0 * scaled-size(1)) + (diff_1 * scaled-size(2)) + (diff_2 * scaled-size(3))
           = (diff_0 * size_1 * size_2 * 1) + (diff_1 * size_2 * 1) + (diff_2 * 1)
           = diff_2 + size_2 * (diff_1 + size_1 * (diff_0))
```

As can be seen, this drastically reduces the number of required multiplication operations and should thus be much faster. It's easy to then generalize this to arbitrary values for the `sizeof...(_Rest)`, which is precisely what `_Distance_from_impl()` does.

Reply:
Thanks! I've added your *`distance-from`* implementation with extra overflow checks!